lib/service/service.go

```diff
@@ -2539,15 +2542,17 @@ func (process *TeleportProcess) setupProxyListeners() (*proxyListeners, error) {
 			Listener:            listener,
 			DisableTLS:          false,
 			DisableSSH:          true,
-			DisableDB:           cfg.Proxy.DisableDatabaseProxy,
+			DisablePostgres:     cfg.Proxy.DisableDatabaseProxy || useSeparatePostgresListener,
 			ID:                  teleport.Component(teleport.ComponentProxy, "web", process.id),
 		})
 		if err != nil {
 			listener.Close()
 			return nil, trace.Wrap(err)
 		}
 		listeners.web = listeners.mux.TLS()
-		listeners.db.postgres = listeners.mux.DB()
+		if err := process.setPostgresListener(cfg, &listeners, useSeparatePostgresListener); err != nil {
```

Comment:
You're already passing `cfg` into this function so I would get rid of the `useSeparatePostgresListener` parameter and just do `!cfg.Proxy.PostgresAddr.IsEmpty()` inside the function again. Would be a bit cleaner interface.